Fixes for security potential issues reported by Riscure

jbech-linaro opened the pull request:
This patch series address a list of potential security issues and (lack of) hardening as identified by [Riscure](https://www.riscure.com/). More details about the issues themselves will be submitted to https://www.op-tee.org/security-advisories in the coming weeks. 

Note that patches here are already "pre-reviewed" (internally at Linaro) and the reason for that is that all patches have already been circulated to trustworthy stakeholders according to our disclosure plan (see the draft [here](https://optee.readthedocs.io/general/disclousure.html)). Having that said, we encourage OP-TEE maintainers and other users to chime in and do a final review here on GitHub.

I've just rebased it on top of master and all tests in xtest passes.

jbech-linaro commented:
I've seen the Shippable error, I'll have a look at it.

jforissier commented on core/arch/arm/plat-d02/conf.mk:
```diff
@@ -5,6 +5,7 @@ CFG_NUM_THREADS ?= 16
 CFG_CRYPTO_WITH_CE ?= y
 CFG_WITH_STACK_CANARIES ?= y
 CFG_CORE_STACK_PROTECTOR_STRONG ?= y
+CFG_CORE_FORTIFY_SOURCE ?= y
```
This should be enabled in `mk/config.mk` instead, since we want it for all platforms by default.

jbech-linaro replied:
Have a fix for it, will push soon.

jforissier commented on ta/mk/ta_dev_kit.mk:
```diff
@@ -69,11 +69,13 @@ libnames += mbedtls
 libdeps += $(ta-dev-kit-dir$(sm))/lib/libmbedtls.a
 endif
 
+ifeq ($(CFG_CC_OPTIMIZE_FOR_SIZE),y)
```
Shouldn't this be folded into "ta-dev-kit: Introduce CFG_TA_FORTIFY_SOURCE"?

jbech-linaro commented on core/arch/arm/plat-d02/conf.mk:
```diff
@@ -4,6 +4,7 @@ $(call force,CFG_TEE_CORE_NB_CORE,16)
 CFG_NUM_THREADS ?= 16
 CFG_CRYPTO_WITH_CE ?= y
 CFG_WITH_STACK_CANARIES ?= y
+CFG_CORE_STACK_PROTECTOR_STRONG ?= y
```
I can move it, just curious, what is our standpoint here? > 50% of platform having a feature enabled means default to `mk/config.mk`? Or is it only when it is 100%? I checked `CFG_WITH_STACK_CANARIES` when I did this and that seem to have it in all `plat-xyz` files. Sounds like this is something that should be mentioned in our platform porting guidelines.

jforissier replied:
I agree that we are lacking some consistency here ;-)

Generally speaking, if a feature is not platform-specific, I think it deserves to be documented in `mk/config.mk` and set to a sensible value at the same time. If we think that most platform should have it enabled by default, then it is `?= y`. Otherwise, `?= n`. The platform config always takes precedence anyway.

jbech-linaro replied:
Makes sense, I'll update the platform guidelines to also mention a few words about that when doing the optee_docs updates.

jbech-linaro replied:
Fix ready, will push soon.

jbech-linaro commented:
> I've seen the Shippable error, I'll have a look at it.

I've changed the macro to only consider SMALL_PAGE_SIZE, since that is what we check against. Will push a fix soon.

jbech-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -101,6 +101,7 @@ static TEE_Result set_tmem_param(const struct optee_msg_param_tmem *tmem,
 static TEE_Result set_rmem_param(const struct optee_msg_param_rmem *rmem,
 				 struct param_mem *mem)
 {
+	size_t req_size = 0;
```
I started out writing the majority of the patches, but in some cases Jens tweaked them. This might be one of them. I'll have a check and will update-accordingly when I force/squash push.

jbech-linaro replied:
Fixed on patch just pushed.

jbech-linaro commented on core/tee/tee_svc.c:
```diff
@@ -679,42 +679,30 @@ static TEE_Result tee_svc_copy_param(struct tee_ta_session *sess,
  * - or it was the original TA memref: update only the size value.
  */
 static TEE_Result tee_svc_update_out_param(
-		struct tee_ta_session *sess,
```
Same here, but I'm almost sure Jens did this one, I'll update accordingly.

jbech-linaro replied:
Fixed on patch just pushed.

jbech-linaro commented on core/core.mk:
```diff
@@ -49,10 +49,21 @@ endif
 cppflags$(sm)  += -DNDEBUG
 endif
+
+# Hardening configuration
```
You're only referring to the comment, yes, I agree. Will fix, btw this actually challenged my git skills. The comment is already in the correct spot, so I simply cannot add a fixup, since there is nothing to change/commit. As I see my only option is to edit it in-place when doing a `git rebase -i` (which I cannot do until everything has been reviewed). If you or anyone else have something up your sleeve, please let me know.

jforissier replied:
;-) indeed this can only be "fixed" during rebase I suppose.

lorc replied:
You can add two fixups. One to remove comment and other one - add it back, but to the right commit.

jbech-linaro replied:
Ah, brilliant!

jbech-linaro replied:
Fixed on patch just pushed.

jbech-linaro commented:
Pushed fixes for all (fixable) comments so far. I've saved the ones that has to be done at rebase/force push time.

jforissier commented:
LGTM, you may add my:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

...to all the commits that deal with CFG_ flags.

etienne-lms commented:
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for the whole with some minor comments.

lorc commented on core/include/optee_msg.h:
```diff
@@ -201,6 +201,15 @@ struct optee_msg_arg {
+ * OPTEE_MSG_GET_ARG_SIZE without wrapping the result.
+ */
+#define OPTEE_MSG_MAX_NUM_PARAMS	\
+	((SIZE_MAX - sizeof(struct optee_msg_arg)) / \
```
Maybe it is worth to use some more meaningful value? Like, `8` for example.

Problem is that `SIZE_MAX` is platform-dependent. So it can cause interesting problems in aarch32/aarch64 interop. Also, right now size of command buffer is partially capped by `SMALL_PAGE_SIZE`, because of check in  ` map_cmd_buffer()`.

jbech-linaro replied:
This is exactly what I had in mind earlier today when asking Jens about it. As you can see on the "fixup", I did change it to use `SMALL_PAGE_SIZE`. This is still quite big, I think I counted that `OPTEE_MSG_MAX_NUM_PARAMS` will be > `190`. I hardly every think you will have anything really bigger than what you said (8), so right now it's more than enough.

jbech-linaro commented:
> `Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for the whole with some minor comments.

@etienne-lms , I'm confused. I cannot see any comments from you. Is GitHub playing me a trick here?

jenswi-linaro commented on core/include/optee_msg.h:
```diff
@@ -204,10 +204,10 @@ struct optee_msg_arg {
  */
 #define OPTEE_MSG_MAX_NUM_PARAMS	\
-	((SIZE_MAX - sizeof(struct optee_msg_arg)) / \
+	((SMALL_PAGE_SIZE - sizeof(struct optee_msg_arg)) / \
```
I'd prefer using `OPTEE_MSG_NONCONTIG_PAGE_SIZE` instead.

jbech-linaro replied:
Fixed and has been pushed.

jbech-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -110,6 +111,14 @@ static TEE_Result set_rmem_param(const struct optee_msg_param_rmem *rmem,
+	 */
+	if (ADD_OVERFLOW(mem->offs, mem->size, &req_size) ||
+	    mem->mobj->size < req_size)
+		return TEE_ERROR_SECURITY;
```
It's an **or** operation and not an **and** operation. This seems perfectly legit to me. I.e., if one of the two conditions is true, then this shall fail. For an **and** operation I think it would still write it like this, but there I'm more with you regarding the suggestion. Agree?

jbech-linaro commented on core/tee/tee_svc.c:
```diff
@@ -679,42 +679,30 @@ static TEE_Result tee_svc_copy_param(struct tee_ta_session *sess,
-			 * If we called a kernel TA the parameters are in shared
-			 * memory and no copy is needed.
+			 * Memory copy is only needed if there's a temporary
+			 * buffer invovled, tmp_buf_va[n] is only update iff
```
Fixed, will push a patch.

jbech-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -479,20 +479,24 @@ static struct mobj *map_cmd_buffer(paddr_t parg, uint32_t *num_params)
+		goto err;
 
 	*num_params = READ_ONCE(arg->num_params);
+
```
Fixed, will push a patch.

jbech-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -824,6 +824,28 @@ static const struct attr_ops attr_ops[] = {
+	return res;
+}
+
+static TEE_Result put_user_u64(uint64_t *dst, size_t value)
```
Isn't that kind of obvious when seeing the current function name and the parameters? On the other hand I didn't write this and I have no strong opinion.

jbech-linaro commented on core/core.mk:
```diff
@@ -49,6 +49,10 @@ endif
+ifeq ($(CFG_CORE_STACK_PROTECTOR_STRONG),y)
+cflags$(sm)	+= -fstack-protector-strong
+endif
+
```
I added that intentionally to make it easier to see it as a block, but I can of course remove it if you don't like it.

jbech-linaro replied:
Oh, if there are double-empty-lines, then I'll fix it. I could only see a single line separating them here on my side. I'll cross check and fix accordingly, thanks!

jbech-linaro commented:
Pushed a few more fixes.

Waiting for:
- Etienne's follow answers on my replies to his finding.

To-do at rebase/force/squash:
- Add Acked-by tags.
- Fix wrong author vs Signed-Off-By.

jbech-linaro commented:
- Fixed SoB / Author inconsistency.
- Fixed all outstanding review comments.
- Fixed spelling error found by checkpatch / Shippable.
- Added A-B and T-B tags.

So, if IBART and friends are happy .... then ...
Squashed, rebased and tag(s) applied (should be) ready for merge! Thanks!

jbech-linaro commented:
@jforissier , makes sense. I'll do another clean-up / clarification round, stay tuned.

jbech-linaro commented:
Easier said that done trying to trigger the protectors, since I get data abort when trying to put together examples that should trigger the stack-protector for example. I'd expect some output like "`*** stack smashing detected ***` at least before the data-abort.

jbech-linaro commented:
I think I can trigger it successfully now, since without using the flags I get data-aborts, with the flags enabled I get stuck. I've tried to figure out where I am by checking GDB, but so far I haven't been able to make sense out of the address. I'll try to see whether it ends up in `stack_check.c` or not.

jforissier commented:
@jbech-linaro sounds good. Perhaps `__stack_chk_fail()` should `EMSG()` and call `panic()` (ifdef __KERNEL__) or `abort()` (user mode TA). Or is there any reason we could not do that in this context?

jbech-linaro commented:
I've tried setting a breakpoint at `__stack_chk_fail`, but it doesn't seem to hang there. But I can see my data (`foo(AAAA...)`) in the `r2` when breaking, so I suspect it is stuck somewhere in strcpy, and/or "panic" while(1) as a consequence of this.
```bash
foo("AAAAAAA.....");

void foo(char *bar) {
char buffer[16];
strcpy(buffer, bar)
}
```

```
r2 = 0x00414141
...
```
I think I will skip the CFG_-patches and submit them as a separate PR if/when I have some time to spare. They weren't really part of the Riscure report, I just added it, since I thought it'd be useful to catch things.

jbech-linaro commented:
Two things ....
1. I've removed the CFG patches from this PR and pushed that.
2. I've rebuilt QEMU with SMP=1 and now!!! Lo and behold! I'm stuck in .... drum roll ... `__stack_chk_fail() { while (1) }`.

So it seems to work as expected.

Anyhow, as before, if IBART and friends are happy:
Squashed, rebased and tag(s) applied (should be) ready for merge! Thanks!
